modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -4683,3 +4688,150 @@ def compare(self, other, **kwargs):
+                ):
+                    copartition = True
+                else:
+                    data_parts = [p[0] for p in data._partitions]
+                    data_part_lengths = data._get_lengths(data_parts, Axis.ROW_WISE)
+                    if any(
+                        lengths[i] != data_part_lengths[i]
+                        for i in range(len(data_parts))
+                    ):
+                        copartition = True
+                if copartition:
+                    _, list_of_right_parts, joined_index, row_lengths = (
+                        frame._copartition(
+                            Axis.ROW_WISE.value,
+                            data,
+                            how="left",
+                            sort=False,
+                            fill_value=True if i == 0 else None,
+                        )
+                    )
+                    data = frame.__constructor__(
+                        list_of_right_parts[0],
+                        joined_index,
+                        data.columns,
+                        row_lengths,
+                        data.column_widths,
+                    )
+                c[i] = data
+            new_caselist.append(tuple(c))
+
+        def map_data(
+            part_offset,
+            part_len,
+            data,
+            data_offset,
+            fill_value,
+            trivial=[is_trivial_idx],
+        ):
+            if isinstance(data, type(frame)):
+                if part_offset < len(data._partitions):
+                    return data._partitions[part_offset][0]._data
+                else:
+                    return [fill_value] * part_len
+
+            if isinstance(data, pandas.Series):
+                if trivial[0] is None:
+                    trivial[0] = is_trivial_index(frame.index)
+                if trivial[0] and is_trivial_index(data.index):
+                    data = data[data_offset : data_offset + part_len]
+                    diff = part_len - len(data)
+                    if diff > 0:
+                        data = pandas.concat((data, pandas.Series([fill_value] * diff)))
+                    return data
+                return data.reindex(
+                    frame.index[data_offset : data_offset + part_len],
+                    fill_value=fill_value,
+                )
+
+            if callable(data) or not is_list_like(data):
+                return data
+
+            return data[data_offset : data_offset + part_len]
+
+        new_parts = []
+        data_offset = 0
+        for i in range(0, parts_len):
+            part = parts[i]
+            part_len = lengths[i]
+            cases = [
+                tuple(
+                    map_data(i, part_len, d[0], data_offset, d[1])
+                    for d in zip(c, (True, None))
+                )
+                for c in new_caselist
+            ]
+            new_parts.append(
+                part.add_to_apply_calls(
```

Comment:
QueryCompilers should not know about low level details. Why can't we use the Modin DataFrame functional?

Reply:
This implementation rely on the partition internals. Should I move it to the DataFrame class?

Reply:
I'm just wondering, whether we can implement `case_when` using [`PandasDataframe.n_ary_op()`](https://github.com/modin-project/modin/blob/cee9b9819803e029c31961b4cf2c913b9f00b962/modin/core/dataframe/pandas/dataframe/dataframe.py#L3665)? It seems like it does the same thing: copartition + a send kernel with broadcasting other partitions

Reply:
[PandasDataframe.n_ary_op()](https://github.com/modin-project/modin/blob/cee9b9819803e029c31961b4cf2c913b9f00b962/modin/core/dataframe/pandas/dataframe/dataframe.py#L3665) operates on partitions, but the caselist may contain not only data frames but also callables and scalars. Can it handle this case?

Reply:
I don't think it can handle this case in the state it is now. We should either add some logic to this operator or add a new operator, which could handle this case.

Reply:
Would it  be reasonable to move the implementation of case_when() to PandasOnRayDataframe? Currently it works only for the Ray backend.

Reply:
Ok to me.

Reply:
In future we could try to pull some general stuff out of it and put in a new core operator, which could be more generic but not only for case_when.

Reply:
> Would it be reasonable to move the implementation of case_when() to PandasOnRayDataframe? Currently it works only for the Ray backend.

Here is a suggestion on how to fix the code for other engines: https://github.com/modin-project/modin/pull/6972#discussion_r1508734112. If we don’t have strong reasons, then it’s better to implement it for all (main) engines at once.

Reply:
Another reason of moving to the data frame class is that the current implementation uses the data frame internals.